### ISSUE TITLE
bot.own_zone should contain only valid positions

### ISSUE THREAD

**Issue by jni** [ASPP/pelita#688]
Currently it contains all coordinates including walls.

**Comment by Debilski** [comment]
@jni Do you think there is some confusion when it also contains the walls? I’m +/-0 on this (if `own_zone` is a square it is much easier to optimise, if we ever find it to be a bottleneck – we’d just make it a class in that case) but I only ever use `own_zone` to check `if bot.position in bot.own_zone` and never use it for anything else.

What would be a reason to use it differently?

**Comment by jni** [comment]
@Debilski I have no memory of raising this, =P I guess some students were running into this. imho own zone could be a set(), and then it's trivially easy and super fast to check whether a position is in there. I don't think you're going to do much faster than that, even with a "rectangle" class.

**Comment by Debilski** [comment]
> @Debilski I have no memory of raising this, =P I guess some students were running into this.

Haha. Ok, I think in future ASPPs we should encourage the students more to open their own issues instead of doing it for them. (I’m guilty of this myself.) Otherwise we’ll never find out how important it really is.

> imho own zone could be a set(), and then it's trivially easy and super fast to check whether a position is in there. I don't think you're going to do much faster than that, even with a "rectangle" class.

Yeah, good point.

**Comment by fmaussion** [comment]
Yes - I am moving from the old OO players to the new functional API, and this is the first thing I came across, Quickly I needed to do `state['homezone'] = set(bot.homezone).difference(set(bot.walls))`, i.e. I think `bot.homezone` should not contain walls.

**Comment by Debilski** [comment]
I sometimes wonder, if we should mention walls at all, and instead only ever mention the empty positions. (OTOH, the list of walls is somewhat easier to explain.)
We are probably going to change bot.homezone and bot.walls to sets as well, which would help with the code either way, otherwise I am rather indifferent about homezone. (Historically, it was just a wrapper around a range and not an explicit list/set, therefore it had to be square and include all positions.)

**Comment by Debilski** [comment]
Setting a milestone for Post-ASPP-2021. The current state is that `bot.homezone` is the set of all positions on one’s side, which is indeed kind of useless, as this can be implemented rather trivially with the expression `0 <= pos[0] < width // 2`.
It seems sensible to me then to include only valid positions in the homezone sets (even though I am not thoroughly convinced that *that* solution would be needed either), but any decision will be postponed. :)
